tests/test_multipart.py

```diff
@@ -146,38 +158,125 @@ def test_multipart_encode_files_allows_filenames_as_none():
     ],
 )
 def test_multipart_encode_files_guesses_correct_content_type(
-    file_name, expected_content_type
-):
+    file_name: str, expected_content_type: str
+) -> None:
     files = {"file": (file_name, io.BytesIO(b"<file content>"))}
     with mock.patch("os.urandom", return_value=os.urandom(16)):
         boundary = binascii.hexlify(os.urandom(16)).decode("ascii")
 
-        stream = encode(data={}, files=files)
+        stream = cast(MultipartStream, encode(data={}, files=files))
+        assert stream.can_replay()
 
         assert stream.content_type == f"multipart/form-data; boundary={boundary}"
-        assert stream.body == (
+        assert b"".join(stream) == (
             f'--{boundary}\r\nContent-Disposition: form-data; name="file"; '
             f'filename="{file_name}"\r\nContent-Type: '
             f"{expected_content_type}\r\n\r\n<file content>\r\n--{boundary}--\r\n"
             "".encode("ascii")
         )
 
 
-def test_multipart_encode_files_allows_str_content():
+def test_multipart_encode_files_allows_str_content() -> None:
     files = {"file": ("test.txt", "<string content>", "text/plain")}
     with mock.patch("os.urandom", return_value=os.urandom(16)):
         boundary = binascii.hexlify(os.urandom(16)).decode("ascii")
 
-        stream = encode(data={}, files=files)
+        stream = cast(MultipartStream, encode(data={}, files=files))
+        assert stream.can_replay()
 
         assert stream.content_type == f"multipart/form-data; boundary={boundary}"
-        assert stream.body == (
+        content = (
             '--{0}\r\nContent-Disposition: form-data; name="file"; '
             'filename="test.txt"\r\n'
             "Content-Type: text/plain\r\n\r\n<string content>\r\n"
             "--{0}--\r\n"
             "".format(boundary).encode("ascii")
         )
+        assert stream.get_headers()["Content-Length"] == str(len(content))
+        assert b"".join(stream) == content
+
+
+def test_multipart_encode_non_seekable_filelike() -> None:
+    """
+    Test that special readable but non-seekable filelike objects are supported,
```

Comment:
Sorry if I missed something in the conversation but could you give an example of such case?

Reply:
This araised more as a theoretical aspect (a branch of the code that'd otherwise result in an exception being raised), but I've seen people being in this situation, eg on SO: [Stream a non-seekable file-like object to multiple sinks](https://stackoverflow.com/questions/39324151/stream-a-non-seekable-file-like-object-to-multiple-sinks) (unfortunately OP doesn't clarify what their non-seekable file-like object looks like). Basically this is a special edge case that theoretically we have to handle, so here we are. :-)